File: verb.cpp
#include <iostream>
#include <string>
#include <vector>
#include <map>

class reg_verb {
    public:

        std::vector<std::string> present;
        std::vector<std::string> imperf;
        std::vector<std::string> pret;
        std::vector<std::string> fut;
        std::vector<std::string> fut_past;

        std::vector<std::string> subj_pres;
        std::vector<std::string> subj_imp;
        std::vector<std::string> subj_fut;
        
        std::vector<std::string> imperative;
        std::string gerund;
        std::string participle;
        std::string infinitive;

        std::string ending;
        std::string stem;

        reg_verb(std::string infin) {
            ending =  infin.substr(infin.length()-2, 2);
            stem = infin.substr(0, infin.length()-2);
            infinitive = infin;
            if (ending == "ar") {
                present.push_back(stem + "o"); present.push_back(stem + "a"); present.push_back(stem + "amos"); present.push_back(stem + "am");
                imperf.push_back(stem + "ava"); imperf.push_back(stem + "ava"); imperf.push_back(stem + "ávamos"); imperf.push_back(stem + "avam");
                pret.push_back(stem + "ei"); pret.push_back(stem + "ou"); pret.push_back(stem + "amos"); pret.push_back(stem + "aram");
                fut.push_back(infin + "ei"); fut.push_back(infin + "á"); fut.push_back(infin + "emos"); fut.push_back(infin + "ão");
                //fut_past
                subj_pres.push_back(stem + "e"); subj_pres.push_back(stem + "e"); subj_pres.push_back(stem + "emos"); subj_pres.push_back(stem + "em");
                subj_imp.push_back(stem + "asse"); subj_imp.push_back(stem + "asse"); subj_imp.push_back(stem + "ássemos"); subj_imp.push_back(stem + "assem");
                subj_fut.push_back(stem + "ar"); subj_fut.push_back(stem + "ar"); subj_fut.push_back(stem + "armos"); subj_fut.push_back(stem + "arem");
                //imperative
                gerund = stem + "ando";
                participle = stem + "ado";
            }
        }

        void print_persons(std::vector<std::string> forms, std::string tense) {
            std::map<int, std::string> person_ref{ {0, "1 sg."}, {1, "3 sg."}, {2, "1 pl."}, {3, "2 pl."} };
            int person_index = 0;
            for (std::vector<std::string>::iterator it = forms.begin(); it != forms.end(); ++it) {
                std::cout << tense << " " << person_ref[person_index] << ": " << *it << "\n";
                person_index++;
            }
            std::cout << "\n";
        }

        void print_forms() {
            std::cout << "\n";
            std::cout << "infinitive: " << infinitive << "\n";
            std::cout << "stem: " << stem  << "-\n";
            std::cout << "ending: -" << ending  << "\n\n";
            print_persons(present, "present");
            print_persons(imperf, "imperfect");
            print_persons(pret, "preterite");
            print_persons(fut, "future");
            print_persons(subj_pres, "present subj.");
            print_persons(subj_imp, "imperfect subj.");
            print_persons(subj_fut, "future subj.");
            std::cout << "gerund: " << gerund << "\n";
            std::cout << "participle: " << participle << "\n";
            std::cout << "\n";
        }
};

int main() {
    reg_verb test("falar");
    test.print_forms();
}

File: java/src/com/Verb.java
package com;

import java.util.HashMap;
import java.util.Map;

public class Verb {
    String infinitive;
    String stem;
    String ending;

    String[] pres = new String[4];
    String[] imperf = new String[4];
    String[] pret = new String[4];
    String[] fut = new String[4];
    String[] fut_past;

    String[] subj_pres = new String[4];
    String[] subj_imp = new String[4];
    String[] subj_fut = new String[4];

    String[] imperative;
    String gerund;
    String participle;

    public Verb(String infin) {
        infinitive = infin;
        stem = infin.substring(0, infin.length()-2);
        ending = infin.substring(infin.length()-2);

        if (ending.equals("ar")) {
            String[] pres_endings = {"o", "a", "amos", "am"};
            String[] imperf_endings = {"ava", "ava", "ávamos", "avam"};
            String[] pret_endings = {"ei", "ou", "amos", "aram"};
            String[] fut_endings = {"ei", "á", "emos", "ão"};
            String[] subj_pres_endings = {"e", "e", "emos", "em"};
            String[] subj_imp_endings = {"asse", "asse", "ássemos", "assem"};
            String[] subj_fut_endings = {"ar", "ar", "armos", "arem"};

            for (int i = 0; i < 4; i++) {
                pres[i] = stem + pres_endings[i];
                imperf[i] = stem + imperf_endings[i];
                pret[i] = stem + pret_endings[i];
                fut[i] = infinitive + fut_endings[i];
                subj_pres[i] = stem + subj_pres_endings[i];
                subj_imp[i] = stem + subj_imp_endings[i];
                subj_fut[i] = stem + subj_fut_endings[i];
            }
        }
    }

    public void print_tense(String[] forms, String name) {
        Map<Integer, String> person_ref = new HashMap<Integer, String>();
        person_ref.put(0, "1 sg."); person_ref.put(1, "3 sg."); person_ref.put(2, "1 pl."); person_ref.put(3, "3 pl.");
        int person_index = 0;
        for (int i = 0; i < 4; i++) {
            System.out.println(name + " " + person_ref.get(person_index) + ": " + forms[person_index]);
            person_index++;
        }
        System.out.println("\n");
    }
    
    public void print_all() {
        System.out.println("infinitive: " + infinitive);
        System.out.println("stem: " + stem);
        System.out.println("ending: " + ending);
        System.out.println("\n");
        print_tense(pres, "present");
        print_tense(imperf, "imperfect");
        print_tense(pret, "preterite");
        print_tense(fut, "future");
        print_tense(subj_pres, "present subj.");
        print_tense(subj_imp, "imperfect subj.");
        print_tense(subj_fut, "future subj.");
    }
}
